### DISCUSSION TITLE
Price template: Tauron G12W + Tauron G12W

### DISCUSSION THREAD

**Discussion by KrzysztofHajdamowicz** [PiotrMachowski/Home-Assistant-custom-components-TGE#9] [Price templates]
Your electricity invoice contains two things: Price of electricity and price of distribution.
One is money you left in power plant for making energy, other one you left to your grid operator for connecting power plant to your utility meter.

To fully maximize price reduction, it's wise to combine dynamic prices with distribution tariff that are cheaper near peak renewables generation, let's say it's G12W.

I'm showing you template to be used as Fixing 1 price to factor prices of G12W distribution tariff into resulting price.
You need to prepare two input_number entities that return price gr/kWh for delivering electrons to your home.

1. Create input_numbers for peak and off-peak distribution rates (For example, Tauron in G12W uses: 0,36898 PLN/kWh as peak and 0,09028 PLN/kWh off-peak during 2024H2)
2. Create `input_number` for comission that Tauron takes for access for dynamic tariffs (~0,11 PLN/kWh if I remember correctly)
3. Use this price modifyer template:
```Python
{% set s = {
    "exchange_rate": states('sensor.eur_to_pln_exchange_rate')|float,
    "VAT": 1.23,
    "peak_distribution_rate": states('input_number.dystrybucja_g12w_szczyt')|float,
    "offpeak_distribution_rate": states('input_number.dystrybucja_g12w_pozaszczyt')|float,
    "market_access_fee": states('input_number.tauron_taryfa_dynamiczna_prowizja')|float
} %}

{# Define peak periods for G12W tariff as a list of tuples (start_hour, stop_hour) #}
{% set peaks = [(6, 13), (15, 22)] %}

{# Determine if it's a weekend (offpeak all day) #}
{% set is_weekend = now().weekday() >= 5 %}

{# Check if the current time falls within any peak period #}
{% set is_peak = False %}
{% if not is_weekend %}
    {% for peak in peaks %}
        {% if now().hour >= peak[0] and now().hour < peak[1] %}
            {% set is_peak = True %}
        {% endif %}
    {% endfor %}
{% endif %}

{# Set distribution rate based on peak or offpeak time #}
{% set distribution_rate = s.peak_distribution_rate if is_peak else s.offpeak_distribution_rate %}

{# Final price calculation: convert current_price to PLN, add distribution rate, market access fee, and apply VAT #}
{{ ((fixing1_rate * s.exchange_rate) + distribution_rate + s.market_access_fee) * s.VAT | float }}
```

**Comment by nerodoxus** [comment]
Why you define here exchange_rate? The data acquired from TGE is in PLN.

**Comment by nerodoxus** [comment]
I found a bug in this code. You set is_peak in the loop but this is only local setting. Specifically, the set statement inside the loop only updates is_peak locally within that specific iteration of the loop. It does not persist the change to is_peak across subsequent iterations of the loop.

Solutions is here. You can experiment with static tmp_hour value and see how ns.is_peak is changed over the loop. 

```
{# Define peak periods for G12W tariff as a list of tuples (start_hour, stop_hour) #}
{% set peaks = [(6, 13), (15, 20), (22, 23)] %}

{# Determine if it's a weekend (offpeak all day) #}
{# set is_weekend = now().weekday() >= 5 #}

{% set is_weekend = False %}

{# set tmp_hour = now().hour #}

{% set tmp_hour = 17 %}

{# Check if the current time falls within any peak period #}
{% set ns = namespace(is_peak=False) %}
{% if not is_weekend %}
    {% for peak in peaks %}
        {% if tmp_hour >= peak[0] and tmp_hour < peak[1] %}
            {% set ns.is_peak = True %}
            set peak {{ ns.is_peak }}  range-> {{ peak[0] }} {{ peak[1] }}
        {% endif %}
        loop - {{ ns.is_peak }} for range-> {{ peak[0] }} {{ peak[1] }}
    {% endfor %}
{% endif %}

{{ ns.is_peak }}
{{ is_weekend }}
```

**Comment by nerodoxus** [comment]
It is worth mentioning that **peak_distribution_rate**, **offpeak_distribution_rate**, and **market_access_fee** modifiers in this template must be defined in the unit **PLN/MWh**, regardless of the unit setting in the add-on setup.

**Reply by PiotrMachowski** [reply]
I have added this info to the documentation and config flow.

**Comment by Jordan2872** [comment]
dzień dobry, chciał bym prosić o pomoc bo nie mam pojęcia jak wypełnić wartościami ten szablon cenowy dla tauron, zaznaczam że nie jestem żadnym programistą  itp:/ z góry dziękuję

**Reply by Mice30** [reply]
Wstawiasz kod i sprawdzasz czy działa:
![image](https://github.com/user-attachments/assets/25821cdf-cf85-4830-beb5-ecf69f5e43bf)

Potem możesz wstawić wpis w configuration.yaml i przeładować HA, żeby widzieć encję i używać jej gdziekolwiek chcesz:
![image](https://github.com/user-attachments/assets/9367abea-2311-486e-82f2-fc2861d0759c)
![image](https://github.com/user-attachments/assets/6a2a73d1-ccbb-4900-ab5e-94b04a0f0198)


tutaj cały kod:

```
sensor:
  - platform: template
    sensors:
      dynamic_energy_price:
        friendly_name: "Dynamiczna cena energii"
        unit_of_measurement: "zł/kWh"
        value_template: >
          {% set s = {
              "exchange_rate": states('sensor.eur_to_pln_exchange_rate')|float(default=4.5),
              "VAT": 1.23,
              "peak_distribution_rate": states('input_number.dystrybucja_g12w_szczyt')|float(default=0.2),
              "offpeak_distribution_rate": states('input_number.dystrybucja_g12w_pozaszczyt')|float(default=0.1),
              "market_access_fee": states('input_number.tauron_taryfa_dynamiczna_prowizja')|float(default=0.03)
          } %}
          
          {% set peaks = [(6, 13), (15, 22)] %}
          {% set is_weekend = now().weekday() >= 5 %}
          {% set ns = namespace(is_peak=False) %}
          
          {% if not is_weekend %}
              {% for peak in peaks %}
                  {% if now().hour >= peak[0] and now().hour < peak[1] %}
                      {% set ns.is_peak = True %}
                  {% endif %}
              {% endfor %}
          {% endif %}
          
          {% set distribution_rate = s.peak_distribution_rate if ns.is_peak else s.offpeak_distribution_rate %}
          {% set fixing1_rate = states('sensor.fixing1_rate')|float(default=0) %}
          
          {{ ((fixing1_rate * s.exchange_rate) + distribution_rate + s.market_access_fee) * s.VAT | float }}


```

**Comment by K80306** [comment]
@KrzysztofHajdamowicz I see there is now() function used within this template. This makes me think price modifier will be applied based on current timestamp. This is correct "now" (at the time of processing/fetching prices), but electricity prices are provided in advance. As a result peak/off peak modifier will be incorrect when rate changes. Do you see such behavior or am I missing something?

**Reply by PiotrMachowski** [reply]
`now` function is overwritten with a one that returns the corrected timestamp that matches the time of data

**Reply by K80306** [reply]
@PiotrMachowski Thank you. Does it mean I do have to write price correction template in config window and should not reference result of "external" template number? I do use such reference now. Could it be a problem?

I do use your integration in pair with https://github.com/jonasbkarlsson/ev_smart_charging/

**Reply by PiotrMachowski** [reply]
I'm not sure if I understand you, can you provide some more specific examples?

**Reply by K80306** [reply]
@PiotrMachowski Sure.

Here are my helpers: (variation of what we have above)
1. `number.electricitynetworkrate` This one is returning peak/off peak network rate
    ```
    {% set minutes_since_midnight = (now().hour * 60) + now().minute %}
    {% if ((minutes_since_midnight > 6* 60 and minutes_since_midnight < 13 * 60) or (minutes_since_midnight > 15* 60 and minutes_since_midnight < 22 * 60)) and now().weekday() <5 %} {{327}}
    {% else %} {{52.0}}
    {% endif %}
    ```

2. `number.electricitynetworkqualityrate` one more price component - constant
    ```
    {{32.1}}
    ```

3. `number.electricitytotalnetworkrate` sum of 1 and 2
    ```
    {{states('number.electricitynetworkrate') | float + states('number.electricitynetworkqualityrate') | float }}
    ```

4.  `number.electricityfixing1skladnikcenotworczy` - one more price component - constant
    ```
    {{89.2}}
    ```


Then configuration of your integration is referencing 3 and 4.
```
{{states('number.electricitytotalnetworkrate') | float + states('number.electricityfixing1skladnikcenotworczy') | float + fixing1_rate}}
```

All this is working "ok" for single moment in time (when processed) but not for the future. I think I'm not allowed to do this due to this overriding not taking effect in my helpers (basically not in scope). Did all of this for further optimization (I can reduce my electricity bill by using it when prices are negative, but cant reduce network fees later), and better visualization

<img width="339" alt="image" src="https://github.com/user-attachments/assets/642baa86-b94d-43d9-a317-5cefe8cbc818" />

Thick blue line: helper identical to integration config
orange columns: integration generated prices
thin purple: number.electricitynetworkrate  (1) based G12 electricity price for reference

**Reply by PiotrMachowski** [reply]
Ok, so this approach won't work - it's not really possible to get future values for helpers

**Reply by K80306** [reply]
Thank you!
